cmd/notation/internal/plugin/plugin.go

```diff
@@ -0,0 +1,179 @@
+	}
+	if !strings.HasPrefix(w, "v") {
+		w = "v" + w
+	}
+	if !semver.IsValid(w) {
+		return 0, fmt.Errorf("%s is not a valid semantic version", w)
+	}
+	return semver.Compare(v, w), nil
+}
+
+// ValidateCheckSum returns nil if SHA256 of file at path equals to checkSum.
+func ValidateCheckSum(path string, checkSum string) error {
+	rc, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	dgst, err := digest.FromReader(rc)
+	if err != nil {
+		return err
+	}
+	enc := dgst.Encoded()
+	if enc != strings.ToLower(checkSum) {
+		return fmt.Errorf("plugin checksum does not match user input. Expecting %s", checkSum)
+	}
+	return nil
+}
+
+// DownloadPluginFromURL downloads plugin file from url to a tmp directory
+func DownloadPluginFromURL(ctx context.Context, url string, tmpFile *os.File) error {
+	// Get the data
+	client := getClient(ctx)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status: %s", resp.Status)
+	}
+	// Write the body to file
+	lr := io.LimitReader(resp.Body, maxPluginSourceBytes)
+	_, err = io.Copy(tmpFile, lr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getClient returns an *auth.Client
+func getClient(ctx context.Context) *auth.Client {
+	client := &auth.Client{
+		Cache:    auth.NewCache(),
+		ClientID: "notation",
+	}
+	client.SetUserAgent("notation/" + version.GetVersion())
+	setHttpDebugLog(ctx, client)
+	return client
+}
+
+// setHttpDebugLog sets http debug logger
+func setHttpDebugLog(ctx context.Context, authClient *auth.Client) {
+	if logrusLog, ok := log.GetLogger(ctx).(*logrus.Logger); ok && logrusLog.Level != logrus.DebugLevel {
+		return
+	}
+	if authClient.Client == nil {
+		authClient.Client = http.DefaultClient
+	}
+	if authClient.Client.Transport == nil {
+		authClient.Client.Transport = http.DefaultTransport
+	}
+	authClient.Client.Transport = trace.NewTransport(authClient.Client.Transport)
+}
+
+// ExtractPluginNameFromFileName checks if fileName is a valid plugin file name
+// and gets plugin name from it based on spec: https://github.com/notaryproject/specifications/blob/main/specs/plugin-extensibility.md#installation
+func ExtractPluginNameFromFileName(fileName string) (string, error) {
+	fname := osutil.FileNameWithoutExtension(fileName)
+	if !strings.HasPrefix(fname, proto.Prefix) {
+		return "", notationerrors.ErrorInvalidPluginFileName{Msg: fmt.Sprintf("invalid plugin file name. Plugin file name requires format notation-{plugin-name}, but got %s", fname)}
+	}
+	_, pluginName, found := strings.Cut(fname, "-")
+	if !found {
+		return "", notationerrors.ErrorInvalidPluginFileName{Msg: fmt.Sprintf("invalid plugin file name. Plugin file name requires format notation-{plugin-name}, but got %s", fname)}
+	}
```

Comment:
Since prefix always has `-` (see https://github.com/notaryproject/notation-go/blob/1bc5a3f8c4d834440deeb3dc0f6855403a6d730d/plugin/proto/proto.go#L19), we don't need to check `found` as it is always `true`.